pkg/morph/event/listener.go

```diff
@@ -181,55 +181,70 @@ func (l *listener) listen(ctx context.Context, intError chan<- error) error {
 
 	l.mtx.RUnlock()
 
-	chEvent, err := l.subscriber.SubscribeForNotification(hashes...)
-	if err != nil {
-		return err
-	}
+	notifyCh := make(chan *state.ContainedNotificationEvent)
 
-	l.listenLoop(ctx, chEvent, intError)
+	var blockCh chan *block.Block
+	if len(l.blockHandlers) > 0 {
+		blockCh = make(chan *block.Block)
+	}
 
-	return nil
-}
+	var notaryCh chan *result.NotaryRequestEvent
+	if l.listenNotary {
+		notaryCh = make(chan *result.NotaryRequestEvent)
+	}
 
-func (l *listener) listenLoop(ctx context.Context, chEvent <-chan *state.ContainedNotificationEvent, intErr chan<- error) {
-	var (
-		blockChan <-chan *block.Block
+	go l.listenLoop(ctx, intError, blockCh, notifyCh, notaryCh)
 
-		notaryChan <-chan *result.NotaryRequestEvent
+	if err := l.subscriber.SubscribeForNotification(notifyCh, hashes...); err != nil {
+		if intError != nil {
+			intError <- fmt.Errorf("could not subscribe to regular events: %w", err)
+		} else {
+			l.log.Debug("could not subscribe to regular events",
+				zap.String("error", err.Error()),
+			)
+		}
 
-		err error
-	)
+		return err
+	}
 
 	if len(l.blockHandlers) > 0 {
-		if blockChan, err = l.subscriber.BlockNotifications(); err != nil {
-			if intErr != nil {
-				intErr <- fmt.Errorf("could not open block notifications channel: %w", err)
+		if err := l.subscriber.BlockNotifications(blockCh); err != nil {
+			if intError != nil {
+				intError <- fmt.Errorf("could not subscribe to block events: %w", err)
 			} else {
-				l.log.Debug("could not open block notifications channel",
+				l.log.Debug("could not subscribe to block events",
 					zap.String("error", err.Error()),
 				)
 			}
 
-			return
+			return err
 		}
-	} else {
-		blockChan = make(chan *block.Block)
 	}
 
 	if l.listenNotary {
-		if notaryChan, err = l.subscriber.SubscribeForNotaryRequests(l.notaryMainTXSigner); err != nil {
-			if intErr != nil {
-				intErr <- fmt.Errorf("could not open notary notifications channel: %w", err)
+		if err := l.subscriber.SubscribeForNotaryRequests(notaryCh, l.notaryMainTXSigner); err != nil {
+			if intError != nil {
+				intError <- fmt.Errorf("could not subscribe to notary events: %w", err)
 			} else {
-				l.log.Debug("could not open notary notifications channel",
+				l.log.Debug("could not subscribe to notary events",
 					zap.String("error", err.Error()),
 				)
 			}
 
-			return
+			return err
 		}
 	}
 
+	return nil
+}
+
+func (l *listener) listenLoop(ctx context.Context, intErr chan<- error,
```

Comment:
Do we need to change it's signature? Ideally, only startup code should be affected.

Reply:
it was not `listenLoop` before: it was `subscribeAndListen`. it could be done in some other way but i found it more readable

> Ideally, only startup code should be affected.

that is startup code changes in fact. just listening and subscribing routines shares some channels now (and they _have_ to do that to prevent dead-lock)